infra/Pos/Network/Types.hs

```diff
@@ -1,55 +1,136 @@
 module Pos.Network.Types
     ( NetworkConfig (..)
-    , emptyNetworkConfig
-
-    , NodeId (..)
-
+    , Topology(..)
+    , topologyNodeType
+    , resolveDnsDomains
+    , defaultNetworkConfig
+    , staticallyKnownPeers
+    , initQueue
+      -- * Re-exports
+      -- ** from .DnsDomains
+    , DnsDomains(..)
+    , DNSError
+      -- ** from time-warp
     , NodeType (..)
     , MsgType (..)
     , Origin (..)
-
-    , NodeName (..)
-    , NodeRegion (..)
-    , NodeRoutes (..)
-    , NodeMetadata (..)
+      -- ** other
+    , NodeId (..)
     ) where
 
 import           Universum
+import           Data.IP (IPv4)
 import           Network.Broadcast.OutboundQueue.Types
+import           Network.Broadcast.OutboundQueue (OutboundQ)
+import qualified Network.Broadcast.OutboundQueue as OQ
 import           Node.Internal (NodeId (..))
+import           Pos.Network.DnsDomains (DnsDomains(..), DNSError)
+import           Pos.Network.Yaml (NodeName(..))
+import           Pos.Util.TimeWarp  (addressToNodeId)
+import qualified Pos.Network.DnsDomains as DnsDomains
+import qualified Data.ByteString.Char8  as BS.C8
 
 -- | Information about the network in which a node participates.
 data NetworkConfig = NetworkConfig
-    { ncNodeType       :: !NodeType
-    -- ^ The role that this node plays in the network.
-    , ncClassification :: !(Map NodeId (NodeType, [[NodeId]]))
-    -- ^ Known classifications of peers, along with their routing information.
+    { ncTopology :: !Topology
+      -- ^ Network topology from the point of view of the current node
+    , ncDefaultPort :: !Word16
+      -- ^ Port number to use when translating IP addresses to NodeIds
+    , ncSelfName :: !(Maybe NodeName)
+      -- ^ Our node name (if known)
     }
+  deriving (Show)
 
-emptyNetworkConfig :: NodeType -> NetworkConfig
-emptyNetworkConfig nt = NetworkConfig
-    { ncNodeType       = nt
-    , ncClassification = mempty
+defaultNetworkConfig :: Topology -> NetworkConfig
+defaultNetworkConfig ncTopology = NetworkConfig {
+      ncDefaultPort = 3000
+    , ncSelfName    = Nothing
+    , ..
     }
 
-newtype NodeName = NodeName Text
-    deriving (Show, Ord, Eq)
+-- | Topology of the network, from the point of view of the current node
+data Topology =
+    -- | All peers of the node have been statically configured
+    --
+    -- This is used for core and relay nodes
+    TopologyStatic !NodeType !(Peers NodeId)
 
--- NodeType comes from
--- Network.Broadcast.OutboundQueue.Classification
--- but it should be a cardano-sl thing.
---data NodeType = NodeCore | NodeRelay | NodeEdge
---    deriving (Show, Ord, Eq)
+    -- | We discover our peers through DNS
+    --
+    -- This is used for behind-NAT nodes.
+  | TopologyBehindNAT !DnsDomains
 
-newtype NodeRegion = NodeRegion Text
-    deriving (Show, Ord, Eq)
+    -- | We discover our peers through Kademlia
+    --
+    -- This is used for exchanges.
```

Comment:
Actually exchanges can use behind NAT mode too, it's orthogonal. The `TopologyP2P` is for users that particularly want to use a P2P mode. In future we may encourage users who host nodes on always-on servers (not on their desktop) to use this mode. This might include exchanges but it's actually orthogonal.